src/index.ts

```diff
@@ -411,62 +404,52 @@ export class Upload extends Pumpify {
-        err = wrapError('Could not authenticate request', err);
-        return callback(err, null!, null);
-      }
-
-      request(authorizedReqOpts, (err, resp, body) => {
-        if (err) {
-          return callback(err, resp, body);
-        }
-
-        if (body && body.error) {
-          return callback(body.error, resp, body);
-        }
-
-        const nonSuccess =
-            Math.floor(resp.statusCode / 100) !== 2;  // 200-299 status code
-        if (nonSuccess &&
-            resp.statusCode !== RESUMABLE_INCOMPLETE_STATUS_CODE) {
-          return callback(new Error(body), resp, body);
-        }
+    reqOpts.validateStatus = (status: number) => {
+      return (status >= 200 && status < 300) ||
+          status === RESUMABLE_INCOMPLETE_STATUS_CODE;
+    };
 
-        callback(null, resp, body);
-      });
-    });
+    this.authClient.request(reqOpts).then(
+        r => {
+          return callback(null, r, r.data);
+        },
+        (err: AxiosError) => {
+          const body = err.response ? err.response.data : undefined;
+          const e = (body && body.error) ? body.error : err;
+          return callback(e, err.response, body);
+        });
   }
 
   private getRequestStream(
       reqOpts: RequestOptions, callback: (requestStream: Request) => void) {
     if (this.userProject) {
-      reqOpts.qs = reqOpts.qs || {};
-      reqOpts.qs.userProject = this.userProject;
+      reqOpts.params = reqOpts.params || {};
+      reqOpts.params.userProject = this.userProject;
     }
 
-    this.authClient.authorizeRequest(reqOpts, (err, authorizedReqOpts) => {
-      if (err) {
-        return this.destroy(wrapError('Could not authenticate request', err));
-      }
-
-      const requestStream = request(authorizedReqOpts);
-      requestStream.on('error', this.destroy.bind(this));
-      requestStream.on('response', this.onResponse.bind(this));
-      requestStream.on('complete', (resp) => {
-        const body = resp.body;
-        if (body && body.error) this.destroy(body.error);
-      });
-
-      // this makes the response body come back in the response (weird?)
-      requestStream.callback = () => {};
-
-      callback(requestStream);
-    });
+    this.authClient.authorizeRequest(reqOpts)
+        .then(opts => {
+          const authorizedReqOpts = axiosToRequest(reqOpts);
+          const requestStream = request(authorizedReqOpts);
+          requestStream.on('error', this.destroy.bind(this));
+          requestStream.on('response', this.onResponse.bind(this));
+          requestStream.on('complete', (resp) => {
+            const body = resp.body;
+            if (body && body.error) this.destroy(body.error);
+          });
+
+          // this makes the response body come back in the response (weird?)
+          requestStream.callback = () => {};
+
+          callback(requestStream);
+        })
+        .catch(err => {
+          return this.destroy(wrapError('Could not authenticate request', err));
```

Comment:
I have no idea.  @stephenplusplus ?

Reply:
The user of the stream would get the error. `this.destroy` is a stream method that exists on the instance itself (from `Pumpify`, since `Upload` extends it). Some more info on what the method does here: https://github.com/stephenplusplus/stream-faqs#whats-the-difference-between-abort-close-end-destroy-finish-and-complete